File: www/js/jalali.js
/* تقویم شمسی — پیاده‌سازی مستقل (پورتِ نسخهٔ پایتونیِ تست‌شده) */
(function (global) {
  'use strict';

  var MONTH_NAMES = ['فروردین', 'اردیبهشت', 'خرداد', 'تیر', 'مرداد', 'شهریور',
    'مهر', 'آبان', 'آذر', 'دی', 'بهمن', 'اسفند'];
  var WEEKDAYS = ['دوشنبه', 'سه‌شنبه', 'چهارشنبه', 'پنج‌شنبه', 'جمعه', 'شنبه', 'یکشنبه'];
  var BREAKS = [-61, 9, 38, 199, 426, 686, 756, 818, 1111, 1181, 1210, 1635,
    2060, 2097, 2192, 2262, 2324, 2394, 2456, 3178];
  var MIN_YEAR = 1178, MAX_YEAR = 1634;
  var ANCHOR_JY = 1400, ANCHOR_G = [2021, 3, 21];
  var PERSIAN = '۰۱۲۳۴۵۶۷۸۹';

  function div(a, b) { return Math.floor(a / b); }
  function mod(a, b) { return a - div(a, b) * b; }

  function segment(jy) {
    var jp = BREAKS[0], jump = 0;
    for (var i = 1; i < BREAKS.length; i++) {
      var jm = BREAKS[i];
      jump = jm - jp;
      if (jy < jm) break;
      jp = jm;
    }
    return { jp: jp, jump: jump };
  }

  function isLeap(jy) {
    if (jy < MIN_YEAR || jy > MAX_YEAR) return false;
    var s = segment(jy), n = jy - s.jp, jump = s.jump;
    if (jump - n < 6) n = n - jump + div(jump + 4, 33) * 33;
    var leap = mod(mod(n + 1, 33) - 1, 4);
    if (leap === -1) leap = 4;
    return leap === 0;
  }

  function monthLength(jy, jm) {
    if (jm <= 6) return 31;
    if (jm <= 11) return 30;
    return isLeap(jy) ? 30 : 29;
  }

  function toDate(gy, gm, gd) { return Date.UTC(gy, gm - 1, gd); }
  function fromDate(ms) { var d = new Date(ms); return [d.getUTCFullYear(), d.getUTCMonth() + 1, d.getUTCDate()]; }

  var cache = {};
  cache[ANCHOR_JY] = toDate(ANCHOR_G[0], ANCHOR_G[1], ANCHOR_G[2]);

  function nowruz(jy) {
    if (jy < MIN_YEAR || jy > MAX_YEAR) throw new Error('سال شمسی خارج از بازه: ' + jy);
    if (cache[jy] !== undefined) return cache[jy];
    var best = null;
    Object.keys(cache).forEach(function (k) {
      var y = parseInt(k, 10);
      if (best === null || Math.abs(y - jy) < Math.abs(best - jy)) best = y;
    });
    var year = best, ms = cache[best];
    while (year !== jy) {
      if (jy > year) { ms += (isLeap(year) ? 366 : 365) * 86400000; year++; }
      else { year--; ms -= (isLeap(year) ? 366 : 365) * 86400000; }
      cache[year] = ms;
    }
    return ms;
  }

  function gregorianToJalali(gy, gm, gd) {
    var ms = toDate(gy, gm, gd);
    var jy = Math.min(MAX_YEAR, Math.max(MIN_YEAR, gy - 621));
    while (ms < nowruz(jy) && jy > MIN_YEAR) jy--;
    while (jy < MAX_YEAR && ms >= nowruz(jy + 1)) jy++;
    var doy = Math.round((ms - nowruz(jy)) / 86400000);
    if (doy < 186) return [jy, 1 + div(doy, 31), 1 + mod(doy, 31)];
    doy -= 186;
    return [jy, 7 + div(doy, 30), 1 + mod(doy, 30)];
  }

  function jalaliToGregorian(jy, jm, jd) {
    if (jy < MIN_YEAR || jy > MAX_YEAR || jm < 1 || jm > 12) throw new Error('تاریخ شمسی نامعتبر');
    var maxd = monthLength(jy, jm);
    if (jd < 1 || jd > maxd) throw new Error('روز شمسی نامعتبر');
    var offset = jm <= 7 ? (jm - 1) * 31 : 186 + (jm - 7) * 30;
    return fromDate(nowruz(jy) + (offset + jd - 1) * 86400000);
  }

  function pad(n) { return (n < 10 ? '0' : '') + n; }

  function parse(value) {
    if (!value) return null;
    if (value instanceof Date) return value;
    var m = /^(\d{4})-(\d{2})-(\d{2})(?:[ T](\d{2}):(\d{2}))?/.exec(String(value).trim());
    if (!m) return null;
    return new Date(+m[1], +m[2] - 1, +m[3], +(m[4] || 0), +(m[5] || 0));
  }

  function format(value, withTime) {
    var dt = parse(value);
    if (!dt) return '';
    var j = gregorianToJalali(dt.getFullYear(), dt.getMonth() + 1, dt.getDate());
    var out = j[0] + '/' + pad(j[1]) + '/' + pad(j[2]);
    if (withTime && (dt.getHours() || dt.getMinutes())) {
      out += ' — ' + pad(dt.getHours()) + ':' + pad(dt.getMinutes());
    }
    return out;
  }

  function todayIso() {
    var d = new Date();
    return d.getFullYear() + '-' + pad(d.getMonth() + 1) + '-' + pad(d.getDate());
  }

  function nowIso() {
    var d = new Date();
    return todayIso() + ' ' + pad(d.getHours()) + ':' + pad(d.getMinutes());
  }

  function persianDigits(text) {
    return String(text).replace(/[0-9]/g, function (d) { return PERSIAN[+d]; });
  }

  function weekday(value) {
    var dt = parse(value);
    if (!dt) return '';
    return WEEKDAYS[(dt.getDay() + 6) % 7];
  }

  function monthName(jm) {
    jm = parseInt(jm, 10);
    return jm >= 1 && jm <= 12 ? MONTH_NAMES[jm - 1] : '';
  }

  function addDays(iso, days) {
    var dt = parse(iso);
    if (!dt) return '';
    dt.setDate(dt.getDate() + days);
    return dt.getFullYear() + '-' + pad(dt.getMonth() + 1) + '-' + pad(dt.getDate());
  }

  function daysBetween(startIso, endIso) {
    var a = parse(startIso), b = parse(endIso);
    if (!a || !b) return null;
    return Math.round((Date.UTC(b.getFullYear(), b.getMonth(), b.getDate())
      - Date.UTC(a.getFullYear(), a.getMonth(), a.getDate())) / 86400000);
  }

  function parseJalali(jy, jm, jd) {
    try {
      var g = jalaliToGregorian(parseInt(jy, 10), parseInt(jm, 10), parseInt(jd, 10));
      return g[0] + '-' + pad(g[1]) + '-' + pad(g[2]);
    } catch (e) {
      return '';
    }
  }

  var api = {
    MONTH_NAMES: MONTH_NAMES,
    isLeap: isLeap,
    monthLength: monthLength,
    gregorianToJalali: gregorianToJalali,
    jalaliToGregorian: jalaliToGregorian,
    format: format,
    parse: parse,
    todayIso: todayIso,
    nowIso: nowIso,
    persianDigits: persianDigits,
    weekday: weekday,
    monthName: monthName,
    addDays: addDays,
    daysBetween: daysBetween,
    parseJalali: parseJalali
  };

  if (typeof module !== 'undefined' && module.exports) module.exports = api;
  global.Jalali = api;
})(typeof window !== 'undefined' ? window : globalThis);

File: legacy/kivy/jalali.py
# -*- coding: utf-8 -*-
"""تقویم شمسی (جلالی) — بدون وابستگی خارجی، مناسب برای بسته‌شدن داخل APK.

روش: قاعدهٔ کبیسهٔ دوره‌های ۳۳ساله (منطبق با تقویم رسمی ایران،
صحت‌سنجی‌شده در برابر دو پیاده‌سازی مرجع برای بازهٔ ۱۳۰۰–۱۶۰۰ شمسی)
و لنگرِ تأییدشدهٔ نوروز: ۱ فروردین ۱۴۰۰ = ۲۱ مارس ۲۰۲۱.
تمام محاسبات با datetime انجام می‌شود تا خطای انباشته نداشته باشیم.
"""
import datetime

JALALI_MONTH_NAMES = [
    "فروردین", "اردیبهشت", "خرداد", "تیر", "مرداد", "شهریور",
    "مهر", "آبان", "آذر", "دی", "بهمن", "اسفند",
]
WEEKDAY_NAMES = ["دوشنبه", "سه‌شنبه", "چهارشنبه", "پنج‌شنبه", "جمعه", "شنبه", "یکشنبه"]
PERSIAN_DIGITS = str.maketrans("0123456789", "۰۱۲۳۴۵۶۷۸۹")

# نقاط شکستِ دوره‌ها (سال‌های شمسی)
_BREAKS = [-61, 9, 38, 199, 426, 686, 756, 818, 1111, 1181,
           1210, 1635, 2060, 2097, 2192, 2262, 2324, 2394, 2456, 3178]

MIN_YEAR = 1178
MAX_YEAR = 1634

# لنگرِ تأییدشده: ۱ فروردینِ این سال شمسی برابر این تاریخ میلادی
_ANCHOR_JY = 1400
_ANCHOR_G = datetime.date(2021, 3, 21)

_NOWRUZ_CACHE = {_ANCHOR_JY: _ANCHOR_G}


def _div(a, b):
    """تقسیم با گردکردن به سمت صفر (هم‌رفتار با پیاده‌سازی مرجع)"""
    return int(a / b) if b else 0


def _mod(a, b):
    return a - _div(a, b) * b


def _segment(jy):
    """بازگرداندن (jp, jump): ابتدا و طولِ دوره‌ای که سال در آن است."""
    jp = _BREAKS[0]
    jump = 0
    for i in range(1, len(_BREAKS)):
        jm = _BREAKS[i]
        jump = jm - jp
        if jy < jm:
            break
        jp = jm
    return jp, jump


def is_leap_jalali(jy):
    """آیا سال شمسی کبیسه است (اسفند ۳۰ روزه)؟"""
    if not (MIN_YEAR <= jy <= MAX_YEAR):
        return False
    jp, jump = _segment(jy)
    n = jy - jp
    if jump - n < 6:
        n = n - jump + _div(jump + 4, 33) * 33
    leap = _mod(_mod(n + 1, 33) - 1, 4)
    if leap == -1:
        leap = 4
    return leap == 0


def jalali_year_length(jy):
    return 366 if is_leap_jalali(jy) else 365


def jalali_month_length(jy, jm):
    if not 1 <= jm <= 12:
        return 0
    if jm <= 6:
        return 31
    if jm <= 11:
        return 30
    return 30 if is_leap_jalali(jy) else 29


def nowruz(jy):
    """تاریخ میلادیِ ۱ فروردینِ سال شمسیِ داده‌شده."""
    if not (MIN_YEAR <= jy <= MAX_YEAR):
        raise ValueError("سال شمسی خارج از بازهٔ پشتیبانی‌شده: %s" % jy)
    # نزدیک‌ترین سالِ کشه‌شده
    best = min(_NOWRUZ_CACHE, key=lambda y: abs(y - jy))
    year, date = best, _NOWRUZ_CACHE[best]
    step = 1 if jy > year else -1
    while year != jy:
        if step > 0:
            date += datetime.timedelta(days=jalali_year_length(year))
            year += 1
        else:
            year -= 1
            date -= datetime.timedelta(days=jalali_year_length(year))
        _NOWRUZ_CACHE[year] = date
    return date


def gregorian_to_jalali(gy, gm, gd):
    """(سال، ماه، روز) میلادی → (سال، ماه، روز) شمسی."""
    date = datetime.date(gy, gm, gd)
    jy = gy - 621
    jy = max(MIN_YEAR, min(MAX_YEAR, jy))
    while date < nowruz(jy) and jy > MIN_YEAR:
        jy -= 1
    while jy < MAX_YEAR and date >= nowruz(jy + 1):
        jy += 1
    doy = (date - nowruz(jy)).days
    if doy < 186:
        jm, jd = 1 + doy // 31, 1 + doy % 31
    else:
        doy -= 186
        jm, jd = 7 + doy // 30, 1 + doy % 30
    return jy, jm, jd


def jalali_to_gregorian(jy, jm, jd):
    """(سال، ماه، روز) شمسی → (سال، ماه، روز) میلادی."""
    if not (MIN_YEAR <= jy <= MAX_YEAR) or not (1 <= jm <= 12):
        raise ValueError("تاریخ شمسی نامعتبر: %s/%s/%s" % (jy, jm, jd))
    maxd = jalali_month_length(jy, jm)
    if not (1 <= jd <= maxd):
        raise ValueError("روز شمسی نامعتبر: %s/%s/%s" % (jy, jm, jd))
    if jm <= 7:
        offset = (jm - 1) * 31
    else:
        offset = 186 + (jm - 7) * 30
    date = nowruz(jy) + datetime.timedelta(days=offset + jd - 1)
    return date.year, date.month, date.day


def _coerce(value):
    if value is None or value == "":
        return None
    if isinstance(value, datetime.datetime):
        return value
    if isinstance(value, datetime.date):
        return datetime.datetime(value.year, value.month, value.day)
    if isinstance(value, str):
        txt = value.strip().replace("T", " ")
        for fmt in ("%Y-%m-%d %H:%M:%S", "%Y-%m-%d %H:%M", "%Y-%m-%d"):
            try:
                return datetime.datetime.strptime(txt, fmt)
            except ValueError:
                continue
    return None


def format_jalali(value, with_time=False, default="-"):
    """نمایش شمسیِ یک تاریخ (رشتهٔ ISO یا date/datetime)."""
    dt = _coerce(value)
    if dt is None:
        return "" if value in (None, "") else str(value)
    try:
        jy, jm, jd = gregorian_to_jalali(dt.year, dt.month, dt.day)
    except (ValueError, OverflowError):
        return default
    out = "%s/%s/%s" % (jy, str(jm).zfill(2), str(jd).zfill(2))
    if with_time and (dt.hour or dt.minute):
        out += " — %s:%s" % (str(dt.hour).zfill(2), str(dt.minute).zfill(2))
    return out


def today_iso():
    return datetime.date.today().isoformat()


def now_iso():
    return datetime.datetime.now().strftime("%Y-%m-%d %H:%M")


def today_jalali():
    return format_jalali(datetime.date.today())


def jalali_weekday(value):
    dt = _coerce(value)
    return WEEKDAY_NAMES[dt.weekday()] if dt else ""


def persian_digits(text):
    return str(text).translate(PERSIAN_DIGITS)


def jalali_month_name(jm):
    try:
        jm = int(jm)
    except (TypeError, ValueError):
        return ""
    return JALALI_MONTH_NAMES[jm - 1] if 1 <= jm <= 12 else ""


def add_days_iso(iso_date, days):
    """جمع/تفریق روز روی تاریخ ISO (خروجی ISO یا رشتهٔ خالی)."""
    dt = _coerce(iso_date)
    if dt is None:
        return ""
    return (dt.date() + datetime.timedelta(days=days)).isoformat()


def days_between_iso(start_iso, end_iso):
    """اختلاف روز (منفی یعنی گذشته)."""
    a, b = _coerce(start_iso), _coerce(end_iso)
    if a is None or b is None:
        return None
    return (b.date() - a.date()).days


def parse_jalali(jy, jm, jd):
    """تبدیل تاریخ شمسی به رشتهٔ ISO (میلادی) — برای ورودیِ کاربر."""
    try:
        return datetime.date(*jalali_to_gregorian(int(jy), int(jm), int(jd))).isoformat()
    except (ValueError, TypeError):
        return ""
